### PR TITLE
docs(lite): harden idd-pr-submit-lite D4 CI-wait edge cases

### LINKED ISSUE #1589: docs(lite): harden idd-pr-submit-lite D4 CI-wait edge cases from PR #1577 review
## Summary

PR #1577 (issue #1544, authoring `idd-pr-submit-lite.instructions.md`)
went through many review rounds hardening D4's CI-wait section. A
late review pass on the merged/frozen HEAD (chatgpt-codex-connector)
surfaced several additional D4/D1-D3 edge cases that are real but
narrower than what the PR could keep absorbing without an unbounded
review-fix loop. This issue captures them for a follow-up pass instead
of reopening #1577's already-stabilized file.

All items below are against
`idd-template/.github/instructions/lite/idd-pr-submit-lite.instructions.md`.
Verify each against current file content and `idd-ci.instructions.md`
before fixing -- some may already be covered by later unrelated edits.

## Candidate gaps

1. **Resume after PR-created-but-D3.5-not-run**: when
   `syncRecommendation: "none"` on a resumed session, D1 step 1 sends
   the session straight to D4, skipping D3.5's closing-keyword
   re-verification. If an earlier session crashed between D3 and
   D3.5, the closing keyword may never get verified.
   (github.com/kurone-kito/idd-skill/pull/1577#discussion_r3627067956)
2. **Resume after push-but-no-PR**: D1 step 1's "no open PR" branch
   assumes local `HEAD` already matches the remote push and jumps to
   D3, without checking for un-pushed local commits made after an
   earlier interrupted push.
   (github.com/kurone-kito/idd-skill/pull/1577#discussion_r3627067962)
3. **Timeout basis for present-but-not-started checks**: D4 step 6
   falls back to "this wait's first poll time" and `runningTimeout`
   for a `checks[]` entry with no `startedAt`, but such an entry
   (queued / not yet running) arguably belongs on `generationTimeout`
   instead, same as a check missing from `checks[]` entirely.
   (github.com/kurone-kito/idd-skill/pull/1577#discussion_r3627067965)
4. **No-required-checks fallback never times out**: D4 step 3's
   fallback to `checks[]` for a `no-required-checks` repository routes
   pending/unknown entries to step 6, but step 6 only evaluates
   entries where `required` is true -- so in a repo with no required
   checks, no timeout/rerun/hold condition can ever fire.
   (github.com/kurone-kito/idd-skill/pull/1577#discussion_r3627067968)
5. **`missing` required check can never reach generation-timeout
   hold**: D4 step 6 only iterates entries already present in
   `checks[]`; a required check that has not appeared at all has no
   entry to iterate, so the generation-timeout path may never
   trigger for it.
   (github.com/kurone-kito/idd-skill/pull/1577#discussion_r3627067971)
6. **Bot-gated `action_required` advisory-convergence recovery**: D4
   step 8's exception only recognizes the `pending: false`
   outstanding-review verdict shape, not the separate case where
   GitHub left a bot-triggered advisory-convergence run at
   `action_required` (a known recoverable case documented in the
   standard CI rerun mechanics).
   (github.com/kurone-kito/idd-skill/pull/1577#discussion_r3627067976)
7. **Rerun budget not carried across resumed waits**: D4 step 6's
   "first timeout for this wait" framing resets on every fresh D4
   entry; a resumed session has no way to know a rerun already
   happened in a prior wait, risking repeated reruns instead of the
   required hold once `rerun-once` is exhausted.
   (github.com/kurone-kito/idd-skill/pull/1577#discussion_r3627129082)
8. **Status-context entries are not Actions runs**: D4 step 6's rerun
   command assumes the timed-out `checks[]` entry's `url` is an
   Actions run URL, but a `type: "status-context"` entry's `url`
   points at an external target, not a rerunnable Actions run.
   (github.com/kurone-kito/idd-skill/pull/1577#discussion_r3627129086)
9. **Waiver rerun re-check is too eager**: D4 step 8's waiver path
   reruns `idd-advisory-convergence` once and immediately re-reads
   `requiredChecks.status` a single time; a fresh rerun is
   asynchronous and commonly still `pending` at that instant, which
   can make a valid waiver look ineffective before it has had time to
   settle.
   (github.com/kurone-kito/idd-skill/pull/1577#discussion_r3627129092)
10. **Stale rerun entries can mis-route advisory routing**: D4 step
    5's "all other required checks already success" scan reads the
    raw `checks[]` array, which can retain a stale failed entry for a
    check that has since passed via rerun, causing the scan to miss
    the D4 step 8 advisory-only exception.
    (github.com/kurone-kito/idd-skill/pull/1577#discussion_r3627129093)
11. **Post-rebase `fix-validate` edits may go uncommitted**: D1 step 7
    runs `fix-validate` after the whole rebase completes; if that
    produces file changes, the lite file does not explicitly say to
    stage and commit them before D2's push. Note: the standard
    `idd-pr-submit.instructions.md` has the same ambiguity in its
    equivalent step, so this may be a standard-file gap rather than a
    lite-specific mismatch -- check there first.
    (github.com/kurone-kito/idd-skill/pull/1577#discussion_r3627129096)

## Notes

- Item 11 may belong against the standard file instead of (or in
  addition to) the lite file, since the same ambiguity exists there.
- These are all narrower edge cases on top of an already-hardened D4
  section; none blocked #1577's merge.

### PR DESCRIPTION
**Pull request by kurone-kito**
## Summary

Hardens `idd-pr-submit-lite.instructions.md`'s D4 (Wait for CI) and D1
(resume) logic against 9 of the 11 edge cases catalogued in #1589 from
PR #1577's review, plus one line shared with the standard PR-submit
file per its own "any mismatch is a bug in this file" contract:

- D1 resume: run D3.5's closing-keyword check even when
  `syncRecommendation: "none"` skips straight to D4; detect and push
  unpushed local commits before continuing.
- D4 step 6: measure a queued (no `startedAt`) or missing-required
  check's timeout against `generationTimeout`, not `runningTimeout`,
  matching `idd-ci.instructions.md`'s canonical Polling algorithm.
- D4 step 3's `no-required-checks` fallback now actually times out
  (previously could poll forever).
- D4 step 6 now folds `requiredChecks.missingNames` into its
  evaluation set, so a required check entirely absent from `checks[]`
  can still reach a generation-timeout hold.
- D4 step 8 gained the bot-gated `action_required` recovery path for
  `idd-advisory-convergence` (rerun the existing non-bot
  `pull_request`-triggered run, never the gated bot run).
- D4 step 6's rerun budget is now resolved from the stalled run's own
  `gh run view --json attempt` history via `ci-wait-policy.mjs
  --rerun-count`, not this wait's in-memory state — survives a
  resumed session.
- D4 step 6 distinguishes a Commit-Status (`status-context`) entry
  (not rerunnable via `gh run rerun`) from an Actions run.
- D4 step 8's waiver path now resumes normal step-6 polling after its
  rerun instead of reading status once immediately (an async rerun is
  commonly still `pending` right after it starts).
- `idd-pr-submit.instructions.md` (standard file): the equivalent
  post-rebase fix-validate step now also commits any resulting
  changes, not just runs the check.

**Not in scope** (per the issue's own note): item #10 (stale-rerun
mis-routing) was re-verified against the current file and found not
to be a distinct defect — `ci-wait-state.mjs` is re-invoked fresh on
every poll iteration, so there is no cached array to mis-route on.

## Self-review notes

This went through an unusually long C1-C6 self-review loop — 7 rounds,
6 of which each caught a genuine High-severity correctness defect in
the same D4 timeout/rerun region, several of them exposing that a
"dedupe `checks[]` by latest `checkName` instance" rule was being
independently restated (and drifting) across three separate steps.
Round 7 did two things: consolidated that rule into one shared
paragraph referenced by all three steps instead of restated by each,
and — while doing so — caught that the restated rule's tie-break
("latest by `completedAt`, else `startedAt`") didn't match what
`isNewerCheckInstance` (`src/scripts/protocol-helpers.mts`) actually
does: an entry with no `completedAt` (still in flight) always
outranks a completed one, regardless of `startedAt`. That's now fixed
in the single consolidated statement, verified directly against the
helper's source.

A row-by-row correspondence check against `idd-ci.instructions.md`'s
Interpretation table found every canonical decision route has a
faithful correspondent in D4 (some validly merged, since
`ci-wait-state.mjs`'s own `FAILURE_STATES` already buckets
`failure`/`cancelled`/`timed_out`/`action_required`/`startup_failure`/`stale`
together).

## Byte budget

`bundleBudgets["bundle-pr-submit-lite"].limitBytes` raised from 22000
to 23500 in `audit/sync-manifest.json` (explicit ratchet callout, per
that field's own documented rule) to accommodate the added edge-case
guidance. Current size: 22225 bytes (source) / 22446 bytes (mirror),
~1050 bytes of headroom remaining.

## Test plan

- [x] `pnpm run lint` (biome, dprint, markdownlint, cspell,
      `node --test`, workshop-integrity) — 2532/2532 tests pass
- [x] `node scripts/audit-docs.mjs --check` passes
- [x] `node scripts/sync-docs.mjs --apply` — mirror in sync, no drift
- [x] Rebased cleanly onto current `main`, re-ran the full gate after

Closes #1589

Co-Authored-By: Claude Sonnet 5 <noreply@anthropic.com>

### CODE REVIEW

File: idd-template/.github/instructions/lite/idd-pr-submit-lite.instructions.md [hunk 1]
```diff
@@ -267,6 +276,17 @@ not resume this wait on its own initiative: a fresh D4 entry over the
...
+**Duplicate check instances**: the state helper's `checks[]` array can
+hold more than one entry for the same `checkName` — a rerun leaves the
+earlier instance in place alongside the fresh one, and this raw array
+is not pre-deduped for callers. Before evaluating `checks[]` below
+(steps 3, 5, and 6 all read it), first reduce it to one entry per
+`checkName`: an entry with no `completedAt` (still in flight) always
+outranks one that has completed, regardless of either entry's
+`startedAt` — a live rerun can never be older than the finished run it
+supersedes. Once both entries have completed, the one with the later
+`completedAt` wins.
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply idd-template/.github/instructions/lite/idd-pr-submit-lite.instructions.md:294]: Accepted -- verified against `src/scripts/ci-wait-state.mts`:
`selectLatestCheckEntryPerName` (used for the required-checks rollup)
deliberately dedupes by `checkName` alone, since the required-check
gate itself matches by name alone and a rerun's `workflowName` can
differ from the run it supersedes -- but the raw `checks[]` array is
otherwise keyed by `(checkName, workflowName)` specifically so
independent checks sharing a display name across workflows stay
distinct (`tests/ci-wait-state.test.mts`). Scoped the dedupe rule:
`checkName`-alone for required entries (steps 5/6), `(checkName,
workflowName)` for step 3's non-required fallback. Fixed in 600f585b.

File: .github/instructions/lite/idd-pr-submit-lite.instructions.md [hunk 2]
```diff
@@ -272,6 +281,17 @@ not resume this wait on its own initiative: a fresh D4 entry over the
...
+**Duplicate check instances**: the state helper's `checks[]` array can
+hold more than one entry for the same `checkName` — a rerun leaves the
+earlier instance in place alongside the fresh one, and this raw array
+is not pre-deduped for callers. Before evaluating `checks[]` below
+(steps 3, 5, and 6 all read it), first reduce it to one entry per
+`checkName`: an entry with no `completedAt` (still in flight) always
+outranks one that has completed, regardless of either entry's
+`startedAt` — a live rerun can never be older than the finished run it
+supersedes. Once both entries have completed, the one with the later
+`completedAt` wins.
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply .github/instructions/lite/idd-pr-submit-lite.instructions.md:299]: Accepted -- verified against `src/scripts/ci-wait-state.mts`:
`selectLatestCheckEntryPerName` (used for the required-checks rollup)
deliberately dedupes by `checkName` alone, since the required-check
gate itself matches by name alone and a rerun's `workflowName` can
differ from the run it supersedes -- but the raw `checks[]` array is
otherwise keyed by `(checkName, workflowName)` specifically so
independent checks sharing a display name across workflows stay
distinct (`tests/ci-wait-state.test.mts`). Scoped the dedupe rule:
`checkName`-alone for required entries (steps 5/6), `(checkName,
workflowName)` for step 3's non-required fallback. Fixed in 600f585b.

File: idd-template/.github/instructions/lite/idd-pr-submit-lite.instructions.md [hunk 3]
```diff
@@ -289,57 +309,84 @@ continuation.
...
+   `requiredChecks.missingNames` (a required check with no `checks[]`
+   entry at all); for step 3's fallback, every still-non-`success`
+   entry regardless of `required` (a `no-required-checks` repository
+   marks none `required`). **Determine
+   timeout from server timestamps, not a client clock estimate**: if
+   the entry has a `startedAt`, elapsed is server time minus
+   `startedAt`, timed out past ci-wait-policy's `runningTimeout`. If it
+   has no `startedAt` (queued) or is a `missingNames` entry, elapsed is
+   server time minus this wait's own first poll time, timed out past
+   `generationTimeout` instead — its running-timeout has not begun.
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply idd-template/.github/instructions/lite/idd-pr-submit-lite.instructions.md]: Accepted -- verified `startedAt: String(entry.startedAt ?? '')` in
`ci-wait-state.mts`: the field is always a string, empty when absent,
never actually missing. Reworded "has/has no startedAt" to
non-empty/empty. Fixed in 600f585b.

File: .github/instructions/lite/idd-pr-submit-lite.instructions.md [hunk 4]
```diff
@@ -294,57 +314,84 @@ continuation.
...
+   `requiredChecks.missingNames` (a required check with no `checks[]`
+   entry at all); for step 3's fallback, every still-non-`success`
+   entry regardless of `required` (a `no-required-checks` repository
+   marks none `required`). **Determine
+   timeout from server timestamps, not a client clock estimate**: if
+   the entry has a `startedAt`, elapsed is server time minus
+   `startedAt`, timed out past ci-wait-policy's `runningTimeout`. If it
+   has no `startedAt` (queued) or is a `missingNames` entry, elapsed is
+   server time minus this wait's own first poll time, timed out past
+   `generationTimeout` instead — its running-timeout has not begun.
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply .github/instructions/lite/idd-pr-submit-lite.instructions.md]: Accepted -- verified `startedAt: String(entry.startedAt ?? '')` in
`ci-wait-state.mts`: the field is always a string, empty when absent,
never actually missing. Reworded "has/has no startedAt" to
non-empty/empty. Fixed in 600f585b.

File: idd-template/.github/instructions/lite/idd-pr-submit-lite.instructions.md [hunk 5]
```diff
@@ -289,57 +309,84 @@ continuation.
...
+   **On timeout**: identify the stalled check (`checkName` plus `url`
+   if present). Two cases have no run `gh run rerun` can target — stop
+   per the condition above and post a hold note naming the check for
+   either: the check has no `url` at all (a required check absent from
+   `checks[]` entirely, i.e. a `missingNames` entry), or its `url` is a
+   Commit-Status (`status-context`) entry pointing at an external
+   target rather than an Actions run. Otherwise resolve the rerun
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply idd-template/.github/instructions/lite/idd-pr-submit-lite.instructions.md]: Accepted -- the phrasing implied `url` itself encoded the
status-context distinction; it's the `type` field that does that
(`url` is just where it points, from `targetUrl`/`detailsUrl`).
Reworded to name `type: "status-context"` directly. Fixed in 600f585b.

File: .github/instructions/lite/idd-pr-submit-lite.instructions.md [hunk 6]
```diff
@@ -294,57 +314,84 @@ continuation.
...
+   **On timeout**: identify the stalled check (`checkName` plus `url`
+   if present). Two cases have no run `gh run rerun` can target — stop
+   per the condition above and post a hold note naming the check for
+   either: the check has no `url` at all (a required check absent from
+   `checks[]` entirely, i.e. a `missingNames` entry), or its `url` is a
+   Commit-Status (`status-context`) entry pointing at an external
+   target rather than an Actions run. Otherwise resolve the rerun
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply .github/instructions/lite/idd-pr-submit-lite.instructions.md]: Accepted -- the phrasing implied `url` itself encoded the
status-context distinction; it's the `type` field that does that
(`url` is just where it points, from `targetUrl`/`detailsUrl`).
Reworded to name `type: "status-context"` directly. Fixed in 600f585b.

File: idd-template/.github/instructions/lite/idd-pr-submit-lite.instructions.md [hunk 7]
```diff
@@ -289,57 +309,84 @@ continuation.
...
    not a CI-wait state — it turns green only after E-phase disposition,
    downstream of D4.
+   - **Bot-gated `action_required`**: if instead `idd-advisory-convergence`
+     is stuck at `action_required` from a gated bot-triggered run (for
+     example Copilot's review event) pending approval, this is a
+     stalled run, not a downstream-of-D4 state. A rerun of the gated
+     run itself keeps the original actor's privileges and re-enters
+     `action_required` — instead `gh run rerun` the existing **non-bot**
+     `pull_request`-triggered run for this HEAD, per `rerunPolicy`,
+     then resume step 6's polling for this one check.
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply idd-template/.github/instructions/lite/idd-pr-submit-lite.instructions.md]: Accepted -- confirmed `CiWaitCheckEntry` carries no actor/event field,
so "identify the non-bot pull_request run" wasn't actionable from
`checks[]` alone. Pointed to the existing
`scripts/rerun-advisory-convergence.mjs` helper instead, which
classifies every idd-advisory-convergence instance via the commit
check-runs API and prints the exact rerun-eligible `gh run rerun`
command. Fixed in 600f585b.

File: .github/instructions/lite/idd-pr-submit-lite.instructions.md [hunk 8]
```diff
@@ -294,57 +314,84 @@ continuation.
...
    not a CI-wait state — it turns green only after E-phase disposition,
    downstream of D4.
+   - **Bot-gated `action_required`**: if instead `idd-advisory-convergence`
+     is stuck at `action_required` from a gated bot-triggered run (for
+     example Copilot's review event) pending approval, this is a
+     stalled run, not a downstream-of-D4 state. A rerun of the gated
+     run itself keeps the original actor's privileges and re-enters
+     `action_required` — instead `gh run rerun` the existing **non-bot**
+     `pull_request`-triggered run for this HEAD, per `rerunPolicy`,
+     then resume step 6's polling for this one check.
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply .github/instructions/lite/idd-pr-submit-lite.instructions.md]: Accepted -- confirmed `CiWaitCheckEntry` carries no actor/event field,
so "identify the non-bot pull_request run" wasn't actionable from
`checks[]` alone. Pointed to the existing
`scripts/rerun-advisory-convergence.mjs` helper instead, which
classifies every idd-advisory-convergence instance via the commit
check-runs API and prints the exact rerun-eligible `gh run rerun`
command. Fixed in 600f585b.

**Comment by kurone-kito** [comment]
_claude-code-58b94a4a: review triage snapshot — IDD automation marker. Do not edit._

File: idd-template/.github/instructions/lite/idd-pr-submit-lite.instructions.md [hunk 9]
```diff
@@ -289,57 +315,90 @@ continuation.
...
+   `runningTimeout`. If it is empty (queued) or the entry is a
+   `missingNames` name with no `checks[]` entry at all, elapsed is
+   server time minus this wait's own first poll time, timed out past
+   `generationTimeout` instead — its running-timeout has not begun.
+   **On timeout**: identify the stalled check (`checkName` plus `url`
+   if present). Two cases have no run `gh run rerun` can target — stop
+   per the condition above and post a hold note naming the check for
+   either: it is a `missingNames` name with no `checks[]` entry at all
+   (so no `url` either), or its `checks[]` entry has `type:
+   "status-context"` (a Commit-Status), whose `url` points at an
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply idd-template/.github/instructions/lite/idd-pr-submit-lite.instructions.md]: Accepted -- verified `url: String(entry?.detailsUrl ?? '')` in
`ci-wait-state.mts` (line 362): a check-run entry can have an empty
`url` even when `type` is `check-run`, if the upstream `detailsUrl`
was itself absent. Added it as a third no-rerun-target case alongside
the missingNames and status-context cases. Fixed in 98278505.

File: .github/instructions/lite/idd-pr-submit-lite.instructions.md [hunk 10]
```diff
@@ -294,57 +320,90 @@ continuation.
...
+   `runningTimeout`. If it is empty (queued) or the entry is a
+   `missingNames` name with no `checks[]` entry at all, elapsed is
+   server time minus this wait's own first poll time, timed out past
+   `generationTimeout` instead — its running-timeout has not begun.
+   **On timeout**: identify the stalled check (`checkName` plus `url`
+   if present). Two cases have no run `gh run rerun` can target — stop
+   per the condition above and post a hold note naming the check for
+   either: it is a `missingNames` name with no `checks[]` entry at all
+   (so no `url` either), or its `checks[]` entry has `type:
+   "status-context"` (a Commit-Status), whose `url` points at an
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply .github/instructions/lite/idd-pr-submit-lite.instructions.md]: Accepted -- verified `url: String(entry?.detailsUrl ?? '')` in
`ci-wait-state.mts` (line 362): a check-run entry can have an empty
`url` even when `type` is `check-run`, if the upstream `detailsUrl`
was itself absent. Added it as a third no-rerun-target case alongside
the missingNames and status-context cases. Fixed in 98278505.

**Comment by kurone-kito** [comment]
_claude-code-58b94a4a: review triage snapshot — IDD automation marker. Do not edit._

**Comment by kurone-kito** [comment]
**Rejected** — chatgpt-codex-connector[bot] did not review HEAD 982785053913e1eedad0dc4777319562d4d0cd16 (Codex usage limits for code reviews reached); this is not a completed review (source: #issuecomment-5044423116)

**Comment by kurone-kito** [comment]
**Rejected** — coderabbitai[bot] did not review HEAD 982785053913e1eedad0dc4777319562d4d0cd16 (review limit reached / rate limited); this is not a completed review (source: #issuecomment-5044423818)

**Comment by kurone-kito** [comment]
_claude-code-58b94a4a: review triage snapshot — IDD automation marker. Do not edit._

**Comment by kurone-kito** [comment]
_claude-code-58b94a4a: review triage snapshot — IDD automation marker. Do not edit._

**Comment by kurone-kito** [comment]
_claude-code-58b94a4a: review triage snapshot — IDD automation marker. Do not edit._
